core/src/blocktree.rs

```diff
@@ -357,6 +380,130 @@ impl Blocktree {
+        let mut prev_inserted_blob_datas = HashMap::new();
+
+        for blob in new_blobs.iter() {
+            let blob = blob.borrow();
+            let blob_slot = blob.slot();
+            let parent_slot = blob.parent();
+
+            // Check if we've already inserted the slot metadata for this blob's slot
+            let entry = slot_meta_working_set.entry(blob_slot).or_insert_with(|| {
+                // Store a 2-tuple of the metadata (working copy, backup copy)
+                if let Some(mut meta) = self
+                    .meta(blob_slot)
+                    .expect("Expect database get to succeed")
+                {
+                    // If parent_slot == std::u64::MAX, then this is one of the dummy metadatas inserted
+                    // during the chaining process, see the function find_slot_meta_in_cached_state()
+                    // for details
+                    if meta.parent_slot == std::u64::MAX {
+                        meta.parent_slot = parent_slot;
+                        // Set backup as None so that all the logic for inserting new slots
+                        // still runs, as this placeholder slot is essentially equivalent to
+                        // inserting a new slot
+                        (Rc::new(RefCell::new(meta.clone())), None)
+                    } else {
+                        (Rc::new(RefCell::new(meta.clone())), Some(meta))
+                    }
+                } else {
+                    (
+                        Rc::new(RefCell::new(SlotMeta::new(blob_slot, parent_slot))),
+                        None,
+                    )
+                }
+            });
+
+            let slot_meta = &mut entry.0.borrow_mut();
+
+            // This slot is full, skip the bogus blob
+            if slot_meta.is_full() {
+                continue;
+            }
+
+            // handle erasure meta info
+            let erasure_meta_entry =
+                erasure_meta_working_set
+                    .entry(blob_slot)
+                    .or_insert_with(|| {
+                        let erasure_meta_stored = self
+                            .erasure_meta_cf
+                            .get_by_slot(blob_slot)
+                            .expect("Expect database get to succeed");
+                        match erasure_meta_stored {
+                            Some(e_meta) => e_meta,
+                            None => ErasureMeta {
+                                slot: blob_slot,
+                                counts: vec![],
+                            },
+                        }
+                    });
+
+            let set_index = blob.index() as usize / erasure::NUM_DATA;
+            while erasure_meta_entry.counts.len() <= set_index {
+                erasure_meta_entry.counts.push((0, 0));
+            }
+
+            erasure_meta_entry.counts[set_index].0 |=
+                1 << blob.index() as usize % erasure::NUM_DATA;
```

Comment:
Are we performing these bit operations to avoid having to do a read operation to check if the value already exists on inserts?

Reply:
Yes. currently that check is still done in the `recover` function in erasure but I was planning to change the module and expose more functionality so blocktree could just gather the blobs given an `ErasureMeta` and do the recovery without scanning for missing indices again